state/in-memory/in_memory.go

```diff
@@ -0,0 +1,412 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inmemory
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+
+	"github.com/dapr/kit/logger"
+
+	"github.com/dapr/components-contrib/state"
+)
+
+type inMemStateStoreItem struct {
+	data   []byte
+	etag   *string
+	expire int64
+}
+
+type inMemoryStore struct {
+	items map[string]*inMemStateStoreItem
+	lock  *sync.RWMutex
+	log   logger.Logger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewInMemoryStateStore(logger logger.Logger) state.Store {
+	return &inMemoryStore{
+		items: map[string]*inMemStateStoreItem{},
+		lock:  &sync.RWMutex{},
+		log:   logger,
+	}
+}
+
+func (store *inMemoryStore) Init(metadata state.Metadata) error {
+	store.ctx, store.cancel = context.WithCancel(context.Background())
+	// start a background go routine to clean expired item
+	go store.startCleanThread()
+	return nil
+}
+
+func (store *inMemoryStore) Close() error {
+	if store.cancel != nil {
+		store.cancel()
+	}
+	// release memory reference
+	store.lock.Lock()
+	defer store.lock.Unlock()
+	store.items = map[string]*inMemStateStoreItem{}
+
+	return nil
+}
+
+func (store *inMemoryStore) Ping() error {
+	return nil
+}
+
+func (store *inMemoryStore) Features() []state.Feature {
+	return []state.Feature{state.FeatureETag, state.FeatureTransactional}
+}
+
+func (store *inMemoryStore) Delete(req *state.DeleteRequest) error {
+	// step1: validate parameters
+	if err := store.doDeleteValidateParameters(req); err != nil {
+		return err
+	}
+
+	// step2 and step3 should be protected by write-lock
+	store.lock.Lock()
+	defer store.lock.Unlock()
+
+	// step2: validate etag if needed
+	if err := store.doValidateEtag(req.Key, req.ETag, req.Options.Concurrency); err != nil {
+		return err
+	}
+
+	// step3: do really delete
+	// this operation won't fail
+	store.doDelete(req.Key)
+	return nil
+}
+
+func (store *inMemoryStore) doDeleteValidateParameters(req *state.DeleteRequest) error {
+	return state.CheckRequestOptions(req.Options)
+}
+
+func (store *inMemoryStore) doValidateEtag(key string, etag *string, concurrency string) error {
+	if etag != nil && *etag != "" && concurrency == state.FirstWrite {
+		// For FirstWrite, we need to validate etag before delete
+		item := store.items[key]
+		if item == nil {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf("state not exist or expired for key=%s", key))
+		}
+		if item.etag == nil || *item.etag != *etag {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf(
+				"state etag not match for key=%s: current=%s, expect=%s", key, *item.etag, *etag))
```

Comment:
`if item.etag == nil` this format causes panic due to the usage of `*item.etag`.

Reply:
we already checked nil at L109 :
![image](https://user-images.githubusercontent.com/26001097/167804219-85c9e8fb-0ed1-42fa-801b-f06a7b69e1ef.png)

So it's safe here.

Reply:
What I would say is that `item.etag` is probably nil, not item. And `fmt.Errorf` uses `*item.etag` without check.

Reply:
@Taction Thanks. 
fixed